platform/security/common/src/main/java/ddf/security/samlp/impl/EntityInformation.java

```diff
@@ -138,28 +146,29 @@ public ServiceInfo getAssertionConsumerService(
   }
 
   private ServiceInfo getAssertionConsumerServiceInfoByUrl(AuthnRequest request) {
-    String assertionConsumerServiceURL = request.getAssertionConsumerServiceURL();
-    List<ServiceInfo> infos =
-        assertionConsumerServices
-            .values()
-            .stream()
-            .filter(service -> assertionConsumerServiceURL.equals(service.getUrl()))
-            .collect(Collectors.toList());
-    ServiceInfo si = null;
-    if (!infos.isEmpty()) {
-      si =
-          infos
-              .stream()
-              .filter(service -> service.getBinding().getUri().equals(request.getProtocolBinding()))
-              .findFirst()
-              .orElse(null);
-    }
-    if (si != null) {
-      return si;
-    } else {
-      return new ServiceInfo(
-          assertionConsumerServiceURL, Binding.from(request.getProtocolBinding()), null);
-    }
+    return getACS(s -> request.getAssertionConsumerServiceURL().equals(s.getUrl()));
+  }
+
+  private ServiceInfo getAssertionConsumerServiceInfoByIndex(Integer index) {
+    return getACS(s -> index.equals(s.getIndex()));
+  }
+
+  private ServiceInfo getAssertionConsumerServiceInfoByProtocolBinding(AuthnRequest request) {
+    return getACS(s -> Binding.from(request.getProtocolBinding()).equals(s.getBinding()));
+  }
+
+  private ServiceInfo getACS(Predicate<ServiceInfo> predicate) {
+    return assertionConsumerServices
+        .values()
+        .stream()
+        .filter(predicate)
+        .filter(serviceInfo -> supportedBindings.contains(serviceInfo.getBinding()))
+        .filter(
+            serviceInfo ->
+                CollectionUtils.intersection(IDP_SUPPORTED_BINDINGS, SSO_RESPONSE_BINDINGS)
```

Comment:
✏️ I've changed my mind. This is actually less efficient - it has to create a new collection for one-time use - than just calling two subsequent filters.